src/Datadog.Trace/HttpOverStreams/DatadogHttpHeaderHelper.cs

```diff
@@ -0,0 +1,37 @@
+using System;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Datadog.Trace.HttpOverStreams
+{
+    internal class DatadogHttpHeaderHelper
+    {
+        public const char CarriageReturn = '\r';
+        public static readonly string NewLine = Environment.NewLine;
+        public static readonly int CrLfLength = NewLine.Length;
+
+        public static async Task WriteLeadingHeaders(HttpRequest request, StreamWriter writer)
+        {
+            // optimization opportunity: cache the ascii-encoded bytes of commonly-used headers
+            await writer.WriteAsync($"{request.Verb} {request.Path} HTTP/1.1{NewLine}").ConfigureAwait(false);
+            await writer.WriteAsync($"Host: {request.Host}{NewLine}").ConfigureAwait(false);
+            await writer.WriteAsync($"Accept-Encoding: identity{NewLine}").ConfigureAwait(false);
+            await writer.WriteAsync($"Content-Length: {request.Content.Length ?? 0}{NewLine}").ConfigureAwait(false);
+
+            await writer.WriteAsync($"{AgentHttpHeaderNames.Language}: .NET{NewLine}").ConfigureAwait(false);
+            await writer.WriteAsync($"{AgentHttpHeaderNames.TracerVersion}: {TracerConstants.AssemblyVersion}{NewLine}").ConfigureAwait(false);
+            // don't add automatic instrumentation to requests from datadog code
+            await writer.WriteAsync($"{HttpHeaderNames.TracingEnabled}: false{NewLine}").ConfigureAwait(false);
```

Comment:
Any reason to not create a single string block with all the headers and use one `writer.WriteAsync` call? I guess that will reduce the complexity of the state machine and the number of calls.

Reply:
It's funny you mention that, I have that change locally.